src/uct/sm/mm/mm_xpmem.c

```diff
@@ -0,0 +1,198 @@
+/**
+ * Copyright (c) UT-Battelle, LLC. 2014-2015. ALL RIGHTS RESERVED.
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2015.  ALL RIGHTS RESERVED.
+ * $COPYRIGHT$
+ * $HEADER$
+ */
+
+#include "mm_pd.h"
+#include "mm_iface.h"
+
+#include <ucs/debug/memtrack.h>
+#include <ucs/debug/log.h>
+#include "xpmem.h"
+
+
+static ucs_status_t uct_xpmem_query()
+{
+    int fd, ver;
+
+    fd = open(XPMEM_DEV_PATH, O_RDWR);
+    if (fd < 0) {
+        ucs_debug("Could not open the XPMEM device file at /dev/xpmem: %m. Disabling xpmem resource");
+        return UCS_ERR_UNSUPPORTED;
+    }
+    close(fd);
+
+    ver = xpmem_version();
+    if (ver < 0) {
+        ucs_debug("Failed to query XPMEM version %d, %m", ver);
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    return UCS_OK;
+}
+
+static ucs_status_t uct_xmpem_reg(void *address, size_t size, 
+                                  uct_mm_id_t *mmid_p)
+{
+    xpmem_segid_t segid; /* 64bit ID */
+    const size_t page_size = ucs_get_page_size();
+    void*  addr_aligned   = (void *)ucs_align_down((uintptr_t)address, page_size);
+    off_t  diff = (uintptr_t)address - (uintptr_t)addr_aligned;
+    size_t length_aligned = ucs_align_up(diff + size, page_size);
+
+    ucs_assert_always(address >= addr_aligned);
+
+    segid = xpmem_make(addr_aligned, length_aligned, XPMEM_PERMIT_MODE, (void *)0666);
+    if (segid < 0) {
+        ucs_error("Failed to register memory with xpmem (addr: %p, len: %zu): %m",
+                  address, size);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    *mmid_p = segid;
+
+    ucs_debug("Calling reg for address %p cookie %lu", address, *mmid_p);
+    return UCS_OK;
+}
+
+static ucs_status_t uct_xpmem_dereg(uct_mm_id_t mmid)
+{
+    int rc;
+    xpmem_segid_t segid = (xpmem_segid_t)mmid;
+
+    ucs_debug("Calling dereg for cookie %lu", mmid);
+
+    ucs_assert_always(segid > 0);
+    rc = xpmem_remove(segid);
+    if (rc < 0) {
+        /* No error since there a chance that it already
+         * was released or deregistered */
+        ucs_debug("Failed to de-register memory: %m");
+    }
+
+    return UCS_OK;
+}
+
+static ucs_status_t uct_xpmem_attach(uct_mm_id_t mmid, size_t length, 
+                                     void *rem_address, 
```

Comment:
To calculate the offset. I'm passing one since I assume that fifo buffer is perfectly aligned.